python_training/XO.py

```diff
@@ -0,0 +1,181 @@
+        #   0 1 2
+        print('   ' + ' '.join([str(num) for num in range(self.board_size)]))
+        # Creates a line that separates the board rows (calculated according to the number of slots)
+        # Eventually the line would look like this -------------
+        separator_line = '  ' + '-' * (3 * self.board_size - (self.board_size - 1))
+        for row_number in range(self.board_size):
+            # Create a line template in which printable values will be formatted.
+            line_template = (str(row_number) + ' |' + '{}|' * self.board_size)
+            # Create a list of printable values according to our mapping.
+            printable_values = []
+            for value in self.board[row_number]:
+                if value is not None:
+                    printable_values.append(self.symbols_mapping[value])
+                else:
+                    printable_values.append(" ")
+            # Format the template with the printable values created above, and print the line.
+            print(line_template.format(*printable_values))
+            # Print a separator line if we haven`t reached the last line.
+            if row_number < self.board_size - 1:
+                print(separator_line)
+
+    def take_turn(self):
+        """
+        Performs several actions:
+        - Switch the player that has the next turn
+        - Receive a slot from the player and check if it available
+        - Positions the players symbol on the board
+        """""
+
+        if self.turns_counter > 0:
+            self.current_player = (self.current_player + 1) % 2
+
+        print('OK {}, Its your turn!'.format(self.players[self.current_player]))
+
+        row_index, column_index = self._get_validated_slot_number_input()
+        self.board[row_index][column_index] = self.current_player
+        self.turns_counter += 1
+
+    def has_game_ended(self) -> bool:
+        """
+        Check whether a player has won/the game was tied.
+        """
+
+        if self.turns_counter < (self.board_size * 2 - 1):
+            return False
+
+        if any([self._check_for_winning_sequence(row) for row in self.board]):
+            return True
+
+        columns_list = [[row[column_number] for row in self.board] for column_number in range(self.board_size)]
+        if any([self._check_for_winning_sequence(column) for column in columns_list]):
+            return True
+
+        left_to_right_diagonal = [self.board[index][index] for index in range(self.board_size)]
+        if self._check_for_winning_sequence(left_to_right_diagonal):
+            return True
+
+        right_to_left_diagonal = [self.board[self.board_size - 1 - index][index] for index in range(self.board_size)]
+        if self._check_for_winning_sequence(right_to_left_diagonal):
+            return True
+
+        if self.turns_counter == self.board_size ** 2:
+            print("The game has ended with a tie!")
+            return True
+
+        else:
```

Comment:
The `return False` part is important, but no need for the actual `else:`.
(You won't get here if any `if` has succeeded, so `else` has no meaning.